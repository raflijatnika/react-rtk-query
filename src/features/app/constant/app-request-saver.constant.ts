export enum REQUEST_SAVER {
  // TODO
  TODO_FETCH_LIST_REQUEST = 'TODO_FETCH_LIST_REQUEST',
  TODO_FETCH_DETAIL_REQUEST = 'TODO_FETCH_DETAIL_REQUEST',
  TODO_CREATE_REQUEST = 'TODO_CREATE_REQUEST',
  TODO_UPDATE_REQUEST = 'TODO_UPDATE_REQUEST',
  TODO_DELETE_REQUEST = 'TODO_DELETE_REQUEST'
}
