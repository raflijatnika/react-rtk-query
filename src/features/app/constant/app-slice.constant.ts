// Interfaces
import { IAppSliceInitialState } from '@/features/app/interfaces/app-slice.interface'

export enum APP_SLICE_LOADING {
  BASE = 'isLoading',
  DETAIL = 'isDetailLoading',
  ACTION = 'isActionLoading',
  DOWNLOAD = 'isDownloadLoading'
}

export const APP_SLICE_INITIAL_STATE: IAppSliceInitialState = {
  app_isInitialized: false,
  app_isConnected: true
}
