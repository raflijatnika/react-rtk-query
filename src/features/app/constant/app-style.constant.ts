export enum APP_COLOR {
  DARK = '#06152B',
  WHITE = '#fff'
}

export enum APP_COLOR_LIGHT {
  PRIMARY = '#3A36DB',
  BACKGROUND = '#F1F4FA'
}
