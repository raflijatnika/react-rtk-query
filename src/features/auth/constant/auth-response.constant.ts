// Interfaces
import {
  IAuthResponseLogin,
  IAuthResponseAuthenticatedUser
} from '@/features/auth/interfaces/auth-response.interface'

export const AUTH_RESPONSE_LOGIN: IAuthResponseLogin = {
  token: 'asdf',
  refreshToken: 'asdf'
}

export const AUTH_RESPONSE_AUTHENTICATED_USER: IAuthResponseAuthenticatedUser =
  {
    id: 'asdf',
    name: 'John Doe',
    email: 'johndoe@gmail.com'
  }
