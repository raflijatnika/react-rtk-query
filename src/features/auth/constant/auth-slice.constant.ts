// Interfaces
import { IAuthSliceInitialState } from '@/features/auth/interfaces/auth-slice.interface'

export const AUTH_SLICE_INITIAL_STATE: IAuthSliceInitialState = {
  auth_authenticatedUser: {
    id: null,
    name: null,
    email: null
  },
  auth_token: {
    token: null,
    refreshToken: null
  }
}
