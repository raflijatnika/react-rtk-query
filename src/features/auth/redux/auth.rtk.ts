// Interfaces
import { IAuthAttrsLogin } from '@/features/auth/interfaces/auth-attrs.interface'
import {
  IAuthResponseToken,
  IAuthResponseAuthenticatedUser
} from '@/features/auth/interfaces/auth-response.interface'

// Rtk
import { emptySplitApi } from '@/features/app/redux/app.rtk'

// Mutations
import { auth_SET_TOKEN } from './auth.slice'

export const authApi = emptySplitApi.injectEndpoints({
  endpoints: builder => ({
    auth_register: builder.mutation<
      IAuthResponseAuthenticatedUser,
      IAuthAttrsLogin
    >({
      query: ({ body }) => ({
        url: '/auth/register',
        method: 'POST',
        body
      })
    }),
    auth_login: builder.mutation<IAuthResponseToken['result'], IAuthAttrsLogin>(
      {
        query: ({ body }) => ({
          url: '/auth/login',
          method: 'POST',
          body
        }),
        transformResponse: (response: IAuthResponseToken) => response.result,
        onQueryStarted: async (payload, { dispatch, queryFulfilled }) => {
          try {
            const response = await queryFulfilled

            dispatch(auth_SET_TOKEN(response.data))
          } catch {
            //
          }
        }
      }
    ),
    auth_me: builder.query<IAuthResponseAuthenticatedUser['result'], void>({
      query: () => ({
        url: `/auth/me`
      })
    })
  }),
  overrideExisting: false
})

export const {
  useAuth_registerMutation,
  useAuth_loginMutation,
  useAuth_meQuery
} = authApi
