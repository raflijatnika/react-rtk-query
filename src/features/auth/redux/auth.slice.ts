// Redux Toolkit
import { createSlice, PayloadAction } from '@reduxjs/toolkit'

// Constant
import { AUTH_SLICE_INITIAL_STATE } from '@/features/auth/constant/auth-slice.constant'

// Interfaces
import {
  IAuthToken,
  IAuthAuthenticatedUser
} from '@/features/auth/interfaces/auth.interface'

const auth = createSlice({
  name: 'auth',
  initialState: AUTH_SLICE_INITIAL_STATE,
  reducers: {
    auth_SET_TOKEN: (
      state,
      { payload: { token, refreshToken } }: PayloadAction<IAuthToken>
    ): void => {
      state.auth_token.token = token
      state.auth_token.refreshToken = refreshToken
    },
    auth_SET_AUTHENTICATED_USER: (
      state,
      { payload }: PayloadAction<IAuthAuthenticatedUser>
    ): void => {
      state.auth_authenticatedUser = payload
    },
    auth_LOGOUT: state => {
      state.auth_token = AUTH_SLICE_INITIAL_STATE.auth_token
      state.auth_authenticatedUser =
        AUTH_SLICE_INITIAL_STATE.auth_authenticatedUser
    }
  }
})

// Mutations
export const { auth_SET_TOKEN, auth_SET_AUTHENTICATED_USER, auth_LOGOUT } =
  auth.actions

export default auth.reducer
