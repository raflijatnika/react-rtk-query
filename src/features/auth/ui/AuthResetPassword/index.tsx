// React
import { useCallback } from 'react'

// Components
import { StyledCard, StyledWrapper } from './components'
import {
  AppBaseButton,
  AppBaseFormItem,
  AppBaseInput,
  AppBaseLabel
} from '@/features/app/components'

// Assets
import AppIcon from '@/assets/images/app.png'

// i18n
import { useTranslation } from 'react-i18next'

// Antd
import { Form } from 'antd'

// Constant
import { APP_COLOR_LIGHT } from '@/features/app/constant/app-style.constant'

// React Router DOM
import { useNavigate } from 'react-router-dom'

// Interfaces
import { IAuthAttrsResetPassword } from '@/features/auth/interfaces/auth-attrs.interface'

const AuthResetPassword = () => {
  // Hook
  const { t } = useTranslation()
  const [form] = Form.useForm()
  const navigate = useNavigate()

  /**
   * @description Redirect to login
   *
   * @return {void} void
   */
  const onRedirectToLogin = useCallback((): void => {
    navigate('/auth/login')
  }, [navigate])

  /**
   * @description Submit handler
   *
   * @param {IAuthAttrsResetPassword} form
   *
   * @return {Promise<void>} Promise<void>
   */
  const onSubmit = useCallback(
    async (form: IAuthAttrsResetPassword): Promise<void> => {
      try {
        //
      } catch (_) {
        //
      }
    },
    []
  )

  return (
    <StyledWrapper>
      <StyledCard>
        {/* Header */}
        <div className='flex flex-col justify-center items-center text-center gap-6 mb-6'>
          <img src={AppIcon} width={'90.81px'} height={'56px'} />

          <AppBaseLabel fontWeight={700} fontSize={26.63}>
            {t('app.recover')}
          </AppBaseLabel>
        </div>

        {/* Form */}
        <Form
          form={form}
          onFinish={onSubmit}
          layout='vertical'
          requiredMark={false}
        >
          {/* Email */}
          <AppBaseFormItem
            name='email'
            label={t('app.user.email')}
            rules={[{ required: true }]}
          >
            <AppBaseInput placeholder='example@example.com' />
          </AppBaseFormItem>

          {/* Button */}
          <AppBaseButton
            type='primary'
            htmlType='submit'
            className='mt-8'
            height={50}
            loading={auth_isActionLoading}
            block
          >
            {t('auth.resetPassword')}
          </AppBaseButton>
        </Form>

        {/* Remember Your Password?  */}
        <div className='mt-6 flex items-center text-center justify-center gap-1'>
          <AppBaseLabel fontSize={16}>
            {t('auth.rememberAccount')}?
          </AppBaseLabel>
          <AppBaseLabel
            fontSize={16}
            fontColor={APP_COLOR_LIGHT.PRIMARY}
            className='cursor-pointer'
            onClick={onRedirectToLogin}
            data-testid='login-button'
          >
            {t('auth.login')}
          </AppBaseLabel>
        </div>
      </StyledCard>
    </StyledWrapper>
  )
}

export { AuthResetPassword }
