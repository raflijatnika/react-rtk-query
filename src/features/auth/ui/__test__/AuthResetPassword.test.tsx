// Test
import { act, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import 'jest-styled-components'

// UI
import { AuthResetPassword } from '../AuthResetPassword'

// Plugins
import { testRedux_renderWithProviders } from '@/plugins/redux/test.redux'

// Constant
import { AUTH_SLICE_INITIAL_STATE } from '@/features/auth/constant/auth-slice.constant'

// React Router DOM
import { useNavigate } from 'react-router-dom'

it('AuthResetPassword should render', () => {
  testRedux_renderWithProviders(<AuthResetPassword />, {
    useRouting: true,
    preloadedState: {
      auth: {
        ...AUTH_SLICE_INITIAL_STATE,
        _persist: { version: 1, rehydrated: true }
      }
    }
  })
})

it('AuthResetPassword should have snapshot', () => {
  const { container } = testRedux_renderWithProviders(<AuthResetPassword />, {
    useRouting: true,
    preloadedState: {
      auth: {
        ...AUTH_SLICE_INITIAL_STATE,
        _persist: { version: 1, rehydrated: true }
      }
    }
  })

  expect(container.innerHTML).toMatchSnapshot()
})

it('AuthResetPassword onRedirectToLogin function should defined', async () => {
  await act(() => {
    testRedux_renderWithProviders(<AuthResetPassword />, {
      useRouting: true
    })
  })

  const user = userEvent.setup()

  await user.click(screen.getByTestId('login-button'))

  const navigate = useNavigate()

  expect(navigate).toHaveBeenCalled()
})

it('AuthResetPassword onSubmit function should defined', async () => {
  //
})
