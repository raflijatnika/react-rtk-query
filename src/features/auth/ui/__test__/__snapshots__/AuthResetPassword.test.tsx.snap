// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthResetPassword should have snapshot 1`] = `"<div class=\\"sc-bczRLJ kBsCEE\\"><div class=\\"sc-gsnTZi gqmGps\\"><div class=\\"flex flex-col justify-center items-center text-center gap-6 mb-6\\"><img src=\\"app.png\\" width=\\"90.81px\\" height=\\"56px\\"><label font-weight=\\"700\\" font-size=\\"26.63\\" class=\\"sc-dkzDqf gKWeuv\\">app.recover</label></div><form class=\\"ant-form ant-form-vertical ant-form-hide-required-mark\\"><div class=\\"ant-form-item sc-jSMfEi jFzftq\\"><div theme=\\"[object Object]\\" class=\\"ant-row ant-form-item-row\\"><div class=\\"ant-col ant-form-item-label\\"><label for=\\"email\\" class=\\"ant-form-item-required\\" title=\\"app.user.email\\">app.user.email</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input placeholder=\\"example@example.com\\" id=\\"email\\" class=\\"ant-input sc-gKXOVf ihLyhw\\" type=\\"text\\" value=\\"\\"></div></div></div></div></div><button type=\\"submit\\" class=\\"ant-btn ant-btn-primary ant-btn-block sc-hKMtZM OXJew mt-8\\"><span>auth.resetPassword</span></button></form><div class=\\"mt-6 flex items-center text-center justify-center gap-1\\"><label font-size=\\"16\\" class=\\"sc-dkzDqf dvQrKU\\">auth.rememberAccount?</label><label font-size=\\"16\\" class=\\"sc-dkzDqf ljQRlH cursor-pointer\\" data-testid=\\"login-button\\">auth.login</label></div></div></div>"`;
